pcsx2/Recording/InputRecording.cpp

```diff
@@ -23,31 +23,72 @@
 #include "SaveState.h"
 
 #include "InputRecording.h"
-#include "Recording/RecordingControls.h"
+#include "InputRecordingControls.h"
 
 #include <vector>
 
 
-// Tag and save framecount along with savestate
+// Save or load PCSX2's global frame counter (g_FrameCount) along with each savestate
+//
+// This is to prevent any inaccuracy issues caused by having a different
+// internal emulation frame count than what it was at the beginning of the
+// original recording
 void SaveStateBase::InputRecordingFreeze()
 {
 	FreezeTag("InputRecording");
 	Freeze(g_FrameCount);
 
 #ifndef DISABLE_RECORDING
-	if (g_FrameCount > 0 && IsLoading())
+	// Loading a save-state is an asynchronous task. If we are playing a recording
+	// that starts from a savestate (not power-on) and the starting (pcsx2 internal) frame
+	// marker has not been set (which comes from the save-state), we initialize it.
+	if (g_InputRecording.IsInitialLoad())
 	{
-		g_InputRecordingData.IncrementUndoCount();
+		g_InputRecordingControls.SetFrameCountTracker(g_FrameCount);
+		g_InputRecording.SetStartingFrame(g_FrameCount);
+	}
+	else if (g_InputRecording.IsActive())
+	{
+		// Explicitly set the frame change tracking variable as to not
+		// detect loading a savestate as a frame being drawn
+		g_InputRecordingControls.SetFrameCountTracker(g_FrameCount);
+
+		if (IsLoading())
+		{
+			g_InputRecording.SetFrameCounter(g_FrameCount);
+		}
 	}
 #endif
 }
 
 #ifndef DISABLE_RECORDING
+// Save or load PCSX2's global frame counter (g_FrameCount) along with each full/fast boot
+//
+// This is to prevent any inaccuracy issues caused by having a different
+// internal emulation frame count than what it was at the beginning of the
+// original recording
+void Pcsx2App::RecordingReset()
+{
+	// Booting is an asynchronous task. If we are playing a recording
+	// that starts from power-on and the starting (pcsx2 internal) frame
+	// marker has not been set, we initialize it.
+	if (g_InputRecording.IsInitialLoad())
+	{
+		g_InputRecordingControls.SetFrameCountTracker(0);
+		g_InputRecording.SetStartingFrame(0);
+	}
+	else if (g_InputRecording.IsActive())
+	{
+		// Explicitly set the frame change tracking variable as to not
+		// detect rebooting as a frame being drawn
+		g_InputRecordingControls.SetFrameCountTracker(0);
+		g_InputRecording.SetFrameCounter(0);
+	}
+}
+
```

Comment:
Rather than defining a function in the App.h file, why not just define it as part of `InputRecording` and call it in AppMain.cpp.  The header already is included in that file where this is presently called.